### PR TITLE
oauth2 login signup prototype in server-cloud

### PR DESCRIPTION
**Pull request by rakeshsharma14317**
Added Login and logout setup in the server-cloud

### CODE REVIEW

File: server-cloud/server/auth.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+func (s *Server) user(c echo.Context) error {
+	sess, err := session.Get("secret", c)
+	if err != nil {
+		fmt.Print(err)
+	}
+
+	profile := sess.Values["profile"]
+	// return c.HTML(http.StatusOK, "profile")
+	return c.Render(http.StatusOK, "user.html", profile)
+}
```
Comment (begelundmuller) [inline server-cloud/server/auth.go:128]: This should return the profile directly as a JSON object – not by rendering a template.

**Review by begelundmuller — CHANGES_REQUESTED** [review]
Thank you for this PR, looks like it gets most of the initial auth setup – very helpful. Some extra comments not addressed in the line-by-line reviews:

- We're not going to serve a frontend from the backend server – the frontend is a SPA in `web-cloud`. So this server should just be a RESTful API server. So remove the stuff in `server/web/` and the related templates and endpoints (like /home and `e.Static`).
- Go CI tests are failing – looks like you didn't run `go fmt ./...` before pushing. If you use VS Code, you can set it up to automatically run format on save.
- Probably add a link to https://auth0.com/docs/quickstart/webapp/golang/01-login in the comment before `Authenticator` to give folks reference to where this setup comes from.

File: server-cloud/server/auth.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+		sess, err := session.Get("secret", c)
+		if err != nil {
+			fmt.Print(err)
+		}
+
+		// fmt.Println("ppp inside IsAuth", sess)
+		if sess.Values["profile"] == nil {
+			fmt.Println(" Nill profile")
+			c.Redirect(http.StatusSeeOther, "/")
+		}
```
[the remark this answers is not in the document]
Comment (rakeshsharma14317) [reply server-cloud/server/auth.go:196]: you mean this redirecting to / or home will be handled by frontend?, What i am doing here is once user logged out, and try to access /user again it will be automatically redirected to the /home or login page

File: server-cloud/server/auth.go (hunk 1 above)
Comment (rakeshsharma14317) [reply server-cloud/server/auth.go:128]: Ohk, just added for testing it using html templated in the sample code, will update this 👍

File: server-cloud/server/auth.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+	if err := sess.Save(c.Request(), c.Response()); err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	// Redirect to logged in page.
+	return c.Redirect(http.StatusTemporaryRedirect, "/auth/user")
+}
+
+func (s *Server) logoutCallback(c echo.Context) error {
```
Comment (begelundmuller) [inline server-cloud/server/auth.go:97]: nit: place `logoutCallback` after the `logout` function (not before)

File: server-cloud/server/server.go [hunk 4]
```diff
@@ -21,13 +26,38 @@ import (
...
 	logger *zap.Logger
 	db     database.DB
+	conf   Config
+	auth   *Authenticator
 }
 
-func New(logger *zap.Logger, db database.DB) *Server {
+type TemplateRegistry struct {
+	templates *template.Template
+}
```
Comment (begelundmuller) [inline server-cloud/server/server.go:35]: This shouldn't be needed anymore

File: server-cloud/server/server.go [hunk 5]
```diff
@@ -21,13 +26,38 @@ import (
...
+	Auth0ClientID     string
+	Auth0ClientSecret string
+	Auth0CallbackURL  string
+	Auth0Domain       string
+	SessionsSecret    string
+}
+
+func (t *TemplateRegistry) Render(w io.Writer, name string, data interface{}, c echo.Context) error {
+	return t.templates.ExecuteTemplate(w, name, data)
+}
```
Comment (begelundmuller) [inline server-cloud/server/server.go:47]: Also not needed anymore

File: server-cloud/server/auth.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+	sess, err := session.Get(authSessionName, c)
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	profile := sess.Values["profile"]
+	return c.JSON(http.StatusOK, profile)
+}
+
+func GetAuth(c context.Context, conf Config) (*Authenticator, error) {
```
Comment (begelundmuller) [inline server-cloud/server/auth.go:147]: Call this function `newAuthenticator` instead, and place it right after the `type Authenticator struct {` struct. Make it lowercase since it's not used externally, only inside this package.

In go, the common structure is a) define a struct, b) define "new" functions for creating the struct, c) define the members functions of the struct.

File: server-cloud/server/server.go [hunk 7]
```diff
@@ -91,14 +121,22 @@ func (s *Server) Serve(ctx context.Context, port int) error {
 	p := prometheus.NewPrometheus("echo", nil)
 	p.Use(e)
 
-	// Add other routes here...
+	gob.Register(map[string]interface{}{})
```
Comment (begelundmuller) [inline server-cloud/server/server.go:124]: Why is this line necessary? If it's for cookie serialization, consider serializing as JSON instead.

File: server-cloud/server/auth.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+		return "", err
+	}
+
+	state := base64.StdEncoding.EncodeToString(b)
+	return state, nil
+}
+
+// IsAuthenticated is a middleware that checks if
+// the user has already been authenticated previously.
+func (s *Server) IsAuthenticated(next echo.HandlerFunc) echo.HandlerFunc {
```
Comment (begelundmuller) [inline server-cloud/server/auth.go:183]: This function is internal, so probably better to make it lowercase. And it doesn't have to be a member of `(s *Server)` since it doesn't access the server. I would make it `func authenticatedMiddleware(next ...`

**Review by begelundmuller — CHANGES_REQUESTED** [review]
Highlighted a few small changes to make, then it's ready to merge

File: server-cloud/cmd/main.go [hunk 9]
```diff
@@ -62,8 +67,20 @@ func main() {
...
+		Auth0CallbackURL:  conf.Auth0CallbackURL,
+		Auth0Domain:       conf.Auth0Domain,
+		SessionsSecret:    conf.SessionsSecret,
+	}
+
 	// Init server
-	server := server.New(logger, db)
+	server, err := server.New(logger, db, srvConf)
+	if err != nil {
+		logger.Fatal("error creating authenticator", zap.Error(err))
```
Comment (begelundmuller) [inline server-cloud/cmd/main.go:82]: Since this is returned from server, the message should be "error creating server"

**Review by begelundmuller — APPROVED** [review]
One tiny thing to fix, then you can merge it
